File: TRViS.UITests.Apple/Sources/Shared/OriginalTimetableV4PageObject.swift
// OriginalTimetableV4PageObject.swift
// Page object for the V4 (Next Big) Original Timetable page.
// Mirrors TRViS.UITests/Pages/OriginalTimetableV4PageObject.cs.

import XCTest

class OriginalTimetableV4PageObject {

    private let app: XCUIApplication
    private let base: BaseUITestCase

    init(app: XCUIApplication, base: BaseUITestCase) {
        self.app = app
        self.base = base
    }

    /// Waits up to `timeout` for V4 to reach a rendered state. TrainStripe,
    /// Hero, MiniList, CompactMiniList, or EmptyState satisfies the wait.
    @discardableResult
    func waitForRendered(timeout: TimeInterval = 30) -> Bool {
        let anchors = [
            AutomationIds.OriginalTimetable.V4.trainStripe,
            AutomationIds.OriginalTimetable.V4.hero,
            AutomationIds.OriginalTimetable.V4.miniList,
            AutomationIds.OriginalTimetable.V4.compactMiniList,
            AutomationIds.OriginalTimetable.V4.emptyState,
        ]
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            for id in anchors {
                if let _ = base.waitForElement(id: id, timeout: 0.5) {
                    return true
                }
            }
            Thread.sleep(forTimeInterval: 0.2)
        }
        return false
    }

    func isEmptyStateVisible(timeout: TimeInterval = 5) -> Bool {
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V4.emptyState, timeout: timeout
        ) != nil
    }

    /// True when the persistent TrainStripe header is visible.
    func isTrainStripeVisible(timeout: TimeInterval = 5) -> Bool {
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V4.trainStripe, timeout: timeout
        ) != nil
    }

    /// True when the Hero card is rendered (i.e. a next-arrival row exists).
    func isHeroVisible(timeout: TimeInterval = 5) -> Bool {
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V4.hero, timeout: timeout
        ) != nil
    }

    /// Reads the TrainStripe TrainNumber Label, empty when not in the tree.
    func getTrainNumber() -> String {
        guard let el = base.waitForElement(
            id: AutomationIds.OriginalTimetable.V4.trainStripeTrainNumber, timeout: 10
        ) else { return "" }
        let label = el.label
        if !label.isEmpty { return label }
        if let v = el.value as? String { return v }
        return ""
    }

    /// True when the Hero card's MarkerBadge has a non-zero frame.
    func isHeroMarkerBadgeVisible(timeout: TimeInterval = 5) -> Bool {
        let id = AutomationIds.OriginalTimetable.V4.heroMarkerBadge
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            let el = app.descendants(matching: .any).matching(identifier: id).firstMatch
            if el.exists {
                let frame = el.frame
                if frame.size.width > 0 && frame.size.height > 0 {
                    return true
                }
            }
            Thread.sleep(forTimeInterval: 0.15)
        }
        return false
    }

    /// True when ANY MarkerBadge (V4-prefixed row or Hero) is visible.
    func isAnyMarkerBadgeVisible(timeout: TimeInterval = 5) -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            // V4 has Hero.MarkerBadge AND Row.<id>.MarkerBadge.
            let predicate = NSPredicate(
                format: "(identifier BEGINSWITH 'OriginalTimetable.V4.Row.' OR identifier == %@) AND identifier ENDSWITH '.MarkerBadge'",
                AutomationIds.OriginalTimetable.V4.heroMarkerBadge
            )
            let query = app.descendants(matching: .any).matching(predicate)
            for i in 0..<min(query.count, 10) {
                let el = query.element(boundBy: i)
                if el.exists {
                    let frame = el.frame
                    if frame.size.width > 0 && frame.size.height > 0 {
                        return true
                    }
                }
            }
            Thread.sleep(forTimeInterval: 0.15)
        }
        return false
    }
}

File: TRViS.UITests/Pages/OriginalTimetableV4PageObject.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Appium;
using TRViS.UITests.Infrastructure;

namespace TRViS.UITests.Pages;

/// <summary>
/// Page object for OriginalTimetableV4Page ("Next Big"). Hero card + MiniList
/// layout. The page is "rendered" when the TrainStripe, Hero, MiniList, or
/// EmptyState anchor is reachable (both layout variants share the strip).
/// </summary>
public class OriginalTimetableV4PageObject : PageObject
{
	public OriginalTimetableV4PageObject(AppiumDriver driver) : base(driver) { }

	public AppiumElement TrainStripe
		=> WaitForElement(AutomationIds.OriginalTimetable.V4.TrainStripe, TimeSpan.FromSeconds(30));
	public AppiumElement TrainNumber
		=> WaitForElement(AutomationIds.OriginalTimetable.V4.TrainStripeTrainNumber, TimeSpan.FromSeconds(30));
	public AppiumElement Hero
		=> FindByAutomationId(AutomationIds.OriginalTimetable.V4.Hero);
	public AppiumElement EmptyState
		=> WaitForElement(AutomationIds.OriginalTimetable.V4.EmptyState, TimeSpan.FromSeconds(30));

	/// <summary>
	/// Waits until V4 has finished its first render. Any of TrainStripe / Hero
	/// / MiniList / EmptyState is sufficient evidence the page reached a paint.
	/// </summary>
	public bool WaitForRendered(double timeoutSeconds = 30)
	{
		var deadline = DateTime.UtcNow.AddSeconds(timeoutSeconds);
		Driver.Manage().Timeouts().ImplicitWait = TimeSpan.Zero;
		try
		{
			string[] anchors = new[]
			{
				AutomationIds.OriginalTimetable.V4.Root,
				AutomationIds.OriginalTimetable.V4.TrainStripe,
				AutomationIds.OriginalTimetable.V4.Hero,
				AutomationIds.OriginalTimetable.V4.MiniList,
				AutomationIds.OriginalTimetable.V4.CompactMiniList,
				AutomationIds.OriginalTimetable.V4.EmptyState,
				AutomationIds.OriginalTimetable.V4.CompactEmptyState,
			};
			while (DateTime.UtcNow < deadline)
			{
				foreach (var id in anchors)
				{
					try
					{
						if (Driver.FindElements(AutomationIdLocator(id)).Count > 0)
							return true;
					}
					catch { }
				}
				Thread.Sleep(200);
			}
			return false;
		}
		finally
		{
			Driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
		}
	}

	/// <summary>True when an EmptyState placeholder is visible (tablet OR compact).</summary>
	public bool IsEmptyStateVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V4.EmptyState, timeoutSeconds)
			|| PollDisplayed(AutomationIds.OriginalTimetable.V4.CompactEmptyState, timeoutSeconds);

	/// <summary>True when the persistent TrainStripe header is visible.</summary>
	public bool IsTrainStripeVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V4.TrainStripe, timeoutSeconds);

	/// <summary>True when the Hero card is rendered (i.e. there's a next-arrival row).</summary>
	public bool IsHeroVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V4.Hero, timeoutSeconds);

	/// <summary>True when the tablet layout grid is in the tree.</summary>
	public bool IsTabletLayoutPresent(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V4.TabletGrid, timeoutSeconds);

	/// <summary>True when the compact layout grid is in the tree.</summary>
	public bool IsCompactLayoutPresent(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V4.CompactGrid, timeoutSeconds);

	/// <summary>
	/// Reads the TrainNumber label from the TrainStripe header. Empty when not
	/// present or when no train is active.
	/// </summary>
	public string GetTrainNumber()
	{
		try
		{
			return TrainNumber.Text ?? string.Empty;
		}
		catch (NoSuchElementException)
		{
			return string.Empty;
		}
	}

	/// <summary>True when the Hero card's MarkerBadge is currently visible.</summary>
	public bool IsHeroMarkerBadgeVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V4.HeroMarkerBadge, timeoutSeconds);
}
